File: java/gpt/p127_nr_ver1.java
// Java (Reader/Writer as file-stream analogs)
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class p127_nr_ver1 {
    private static final int MAX_USERNAME_LEN = 64;
    private static final Pattern USER_RE = Pattern.compile("(?:\\b|&)username=([^&\\s\\r\\n]+)");
    private static final Pattern ALLOWED_RE = Pattern.compile("^[A-Za-z0-9_.-]{1,64}$");

    private static void writeResponse(Writer response, int status, String reason, String body) throws IOException {
        byte[] bodyBytes = body.getBytes(StandardCharsets.UTF_8);
        response.write("HTTP/1.1 " + status + " " + reason + "\r\n");
        response.write("Content-Type: text/plain; charset=utf-8\r\n");
        response.write("X-Content-Type-Options: nosniff\r\n");
        response.write("Connection: close\r\n");
        response.write("Content-Length: " + bodyBytes.length + "\r\n");
        response.write("\r\n");
        response.write(body);
        response.flush();
    }

    private static String extractUsername(String raw) {
        Matcher m = USER_RE.matcher(raw);
        return m.find() ? m.group(1) : null;
    }

    private static boolean validateUsername(String u) {
        return u != null && u.length() >= 1 && u.length() <= MAX_USERNAME_LEN && ALLOWED_RE.matcher(u).matches();
    }

    public static void http_hello_page(Reader request, Writer response) {
        if (request == null || response == null) return;

        final int MAX_REQ_CHARS = 1_000_000;
        StringBuilder sb = new StringBuilder(4096);

        try {
            char[] buf = new char[4096];
            int n;
            int total = 0;
            while ((n = request.read(buf)) != -1) {
                total += n;
                if (total > MAX_REQ_CHARS) {
                    writeResponse(response, 413, "Payload Too Large", "Request too large\n");
                    return;
                }
                sb.append(buf, 0, n);
            }

            String raw = sb.toString();
            String username = extractUsername(raw);
            if (username == null) {
                writeResponse(response, 400, "Bad Request", "Bad Request\n");
                return;
            }

            if (!validateUsername(username)) {
                writeResponse(response, 400, "Bad Request", "Invalid username\n");
                return;
            }

            writeResponse(response, 200, "OK", "Hello " + username + "\n");
        } catch (IOException e) {
            try {
                writeResponse(response, 400, "Bad Request", "Bad Request\n");
            } catch (IOException ignored) { }
        }
    }
}

File: python/gpt/p127_nr_ver1.py
# Python (file-like request/response objects)
import re

MAX_USERNAME_LEN = 64
_ALLOWED_RE = re.compile(r"^[A-Za-z0-9_.-]{1,64}$")

def _write_response(response, status: int, reason: str, body: str) -> None:
    body_bytes = body.encode("utf-8", errors="replace")
    headers = (
        f"HTTP/1.1 {status} {reason}\r\n"
        "Content-Type: text/plain; charset=utf-8\r\n"
        "X-Content-Type-Options: nosniff\r\n"
        "Connection: close\r\n"
        f"Content-Length: {len(body_bytes)}\r\n"
        "\r\n"
    ).encode("ascii", errors="strict")
    response.write(headers)
    response.write(body_bytes)

def _extract_username(raw: str) -> str | None:
    m = re.search(r"(?:\b|&)username=([^&\s\r\n]+)", raw)
    return m.group(1) if m else None

def _validate_username(u: str) -> bool:
    return bool(_ALLOWED_RE.fullmatch(u or ""))

def http_hello_page(request, response) -> None:
    try:
        raw_bytes = request.read()
        if isinstance(raw_bytes, str):
            raw = raw_bytes
        else:
            raw = raw_bytes.decode("utf-8", errors="replace")
    except Exception:
        _write_response(response, 400, "Bad Request", "Bad Request\n")
        return

    username = _extract_username(raw)
    if username is None:
        _write_response(response, 400, "Bad Request", "Bad Request\n")
        return

    if not _validate_username(username):
        _write_response(response, 400, "Bad Request", "Invalid username\n")
        return

    _write_response(response, 200, "OK", f"Hello {username}\n")